plasmapy/particles/_factory.py

```diff
@@ -9,15 +9,106 @@
 import astropy.units as u
 import contextlib
 
-from astropy.units.physical import electrical_charge, mass
-from numbers import Integral
-from typing import Union
+from astropy.constants import m_e
+from numbers import Integral, Real
+from typing import Any, Union
 
 from plasmapy.particles.exceptions import InvalidParticleError
 from plasmapy.particles.particle_class import CustomParticle, Particle
 from plasmapy.particles.particle_collections import ParticleList
 
 
+def _generate_error_message(args: tuple, kwargs: dict[str, Any]) -> str:
+    """Compose an error message for invalid particles."""
+
+    errmsg = "Unable to create a particle from: "
+
+    if args:
+        errmsg += repr(args)
+        if kwargs:
+            errmsg += " and "
+
+    if kwargs:
+        errmsg += repr(kwargs)
+
+    errmsg += (
+        ". For information on creating particles, see: "
+        "https://docs.plasmapy.org/en/stable/glossary.html"
+    )
+
+    return errmsg
+
+
+def _make_custom_particle_with_real_charge_number(*args, **kwargs):
+    """
+    Create a |CustomParticle| for mean or composite ions.
+
+    This function is intended to produce |CustomParticle| instances
+    provided a string representing an element or isotope without charge
+    information (e.g., ``"He"`` or ``"He-4"`` but not ``"He-4 2+"``)
+    along with a charge number that is a real number but not an integer.
+
+    Parameters
+    ----------
+    *args : (1,) tuple of str
+        A `tuple` containing a representation of an element or isotope,
+        without charge information.
+
+    *kwargs : dict of str to real numbers
+        Keyword arguments like those that can be passed to |Partile|,
+        except where the charge number ``Z`` is a real number but not
+        an integer.
+
+    Raises
+    ------
+    |InvalidParticleError|
+        If the |CustomParticle| cannot be created.
+
+    Examples
+    --------
+    >>> _make_custom_particle_with_real_charge_number("He-4", Z=1.5)
+    CustomParticle(mass=6.64511...e-27 kg, charge=2.40326...e-19 C)
+    >>> _make_custom_particle_with_real_charge_number("He", Z=1.5, mass_numb=4)
+    CustomParticle(mass=6.64511...e-27 kg, charge=2.40326...e-19 C)
+    """
+
+    if len(args) != 1 or "Z" not in kwargs or not isinstance(args[0], (Integral, str)):
+        raise InvalidParticleError(
+            "Cannot create CustomParticle with this function with "
+            f"{args = } and {kwargs = }."
+        )
+
+    Z = kwargs.pop("Z")
+
+    if not isinstance(Z, (Real, u.Quantity)):
+        raise InvalidParticleError("The charge number must be a real number.")
+
+    base_particle = Particle(*args, **kwargs, Z=0)
+
+    if not base_particle.is_category(require="element", exclude="ion"):
+        # Add a test if this function becomes part of public API
+        raise InvalidParticleError("Cannot create CustomParticle.")  # coverage: ignore
```

Comment:
This line shouldn't get called by anything in the public API, but I put it in here as a failsafe in case we do make this function part of the public API.  I didn't test it directly since testing the private API can make implementation details harder to change.